src/hashset.h

```diff
@@ -0,0 +1,188 @@
+#include <unistd.h>
+
+/* --- Opaque types --- */
+
+typedef struct hashset hashset;
+typedef struct hashsetBucket hashsetBucket;
+typedef struct hashsetStats hashsetStats;
+
+/* --- Non-opaque types --- */
+
+/* The hashsetType is a set of callbacks for a hashset. All callbacks are
+ * optional. With all callbacks omitted, the hashset is effectively a set of
+ * pointer-sized integers. */
+typedef struct {
+    /* If the type of an element is not the same as the type of a key used for
+     * lookup, this callback needs to return the key within an element. */
+    const void *(*elementGetKey)(const void *element);
+    /* Hash function. Defaults to hashing the bits in the pointer, effectively
+     * treating the pointer as an integer. */
+    uint64_t (*hashFunction)(const void *key);
+    /* Compare function, returns 0 if the keys are equal. Defaults to just
+     * comparing the pointers for equality. */
+    int (*keyCompare)(hashset *s, const void *key1, const void *key2);
+    /* Callback to free an element when it's overwritten or deleted.
+     * Optional. */
+    void (*elementDestructor)(hashset *s, void *element);
+    /* Callback to control when resizing should be allowed. */
+    int (*resizeAllowed)(size_t moreMem, double usedRatio);
+    /* Invoked at the start of rehashing. Both tables are already created. */
+    void (*rehashingStarted)(hashset *s);
+    /* Invoked at the end of rehashing. Both tables still exist and are cleaned
+     * up after this callback. */
+    void (*rehashingCompleted)(hashset *s);
+    /* Track memory usage using this callback. It is called with a positive
+     * number when the hashset allocates some memory and with a negative number
+     * when freeing. */
+    void (*trackMemUsage)(hashset *s, ssize_t delta);
+    /* Allow a hashset to carry extra caller-defined metadata. The extra memory
+     * is initialized to 0. */
+    size_t (*getMetadataSize)(void);
+    /* Flag to disable incremental rehashing */
+    unsigned instant_rehashing : 1;
+} hashsetType;
+
+typedef enum {
+    HASHSET_RESIZE_ALLOW = 0,
+    HASHSET_RESIZE_AVOID,
+    HASHSET_RESIZE_FORBID,
+} hashsetResizePolicy;
+
+typedef void (*hashsetScanFunction)(void *privdata, void *element);
+
+/* Constants */
+#define HASHSET_BUCKET_SIZE 64 /* bytes, the most common cache line size */
+
+/* Scan flags */
+#define HASHSET_SCAN_EMIT_REF (1 << 0)
+
+/* Iterator. To be able to use a stack-allocated iterator, it's not opaque. The
+ * fields in this struct should not be accessed though. Just use the hashset
+ * iterator functions. */
+typedef struct {
+    hashset *hashset;
+    hashsetBucket *bucket;
+    long index;
+    int table;
+    int pos_in_bucket;
+    uint64_t safe : 1;
+    union {
+        /* Unsafe iterator fingerprint for misuse detection. */
+        uint64_t fingerprint : 63;
+        /* Safe iterator temporary storage for bucket chain compaction. */
+        uint64_t last_seen_size : 63;
+    };
+} hashsetIterator;
```

Comment:
I was concerned that malloc/free would be a degradation compared to stack allocation, which we have for dict. Do you think its negligible?

A type like `long[20]` or `struct{ char junk[100]; }` is a hack I considered. It would require more casting and more work to make sure we cover all the esoteric platforms we support. To be on the safe side, we would probably need to make it larger than what it actually is.

The iterator contains a `bucket` pointer which is actually an opaque struct. Without the bucket, the caller can't really make any of the internals of the iterator struct. I guess at AWS, you find it really hard to resist abusing abstractions like this? :) In Valkey, I think we would prevent misuse simply by code review.

Reply:
oh, c'mon now - there are plenty of places in Valkey where we are already/historically abusing structs that are exposed.

I don't think that something like `long[20]` is really a hack.  It's an opaque item with a defined size.  I definitely think the size needs to be validated in static assertions (compile time).  Casting is largely eliminated by localizing it in a single typed conversion function.  This lets the compiler ensure that we're casting the right type - so it's safer than a blind cast.

>I guess at AWS, you find it really hard to resist abusing abstractions...

I think that given a large enough sample size, someone will abuse the abstraction if it's possible to do so.  AWS, indeed, provides a large sample size - but I think the open-source community at large is even larger.  IMO it's best to use design principles which avoid presenting such an option.

Reply:
I do prefer opaque structs and I know the dict internals are abused in numerous places within Valkey. The recent file `memory_prefetching.c` for example.

I do want prefer opaque structs, but I do think defining something to `long[20]` that is not an array of 20 longs is a hack. It can also be called a workaround for the fact that C doesn't provide opaque types with a non-opaque size.

I'll look into the `long[20]` hack later. If it doesn't become too ugly, I'll take it. :)

Reply:
Done.